File: python/159_LongestSubstringWithAtMostTwoDistinctCharacters.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        letter_counts = {chr(ord(c) + i): 0 for c in ['a', 'A'] for i in range(26)}
        distincts = max_length = length = trail_idx = 0
        
        for i, c in enumerate(s):
            letter_counts[c] += 1
            length += 1
            if letter_counts[c] == 1:
                distincts += 1
            while distincts > 2:
                letter_counts[s[trail_idx]] -= 1
                length -= 1
                if letter_counts[s[trail_idx]] == 0:
                    distincts -= 1
                trail_idx += 1
            max_length = max(max_length, length)
            
        return max_length

File: java/159_LongestSubstringWithAtMostTwoDistinctCharacters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        HashMap<Character, Integer> letterCounts = new HashMap<Character, Integer>(){{
            char[] startingPoints = {'a', 'A'};
            for (char c : startingPoints) {
                for (int i = 0; i < 26; ++i) {
                    put((char) (c + i), 0);
                }
            }
        }};
        
        int trailIdx = 0, maxLength = 0, currentLength = 0, distinctCharCount = 0;
        for (int i = 0; i < s.length(); ++i) {
            char currChar = s.charAt(i);
            currentLength++;
            letterCounts.put(currChar, letterCounts.get(currChar) + 1);
            if (letterCounts.get(currChar) == 1) distinctCharCount++;
            while (distinctCharCount > 2) {
                char trailChar = s.charAt(trailIdx++);
                letterCounts.put(trailChar, letterCounts.get(trailChar) - 1);
                if (letterCounts.get(trailChar) == 0) distinctCharCount--;
                currentLength--;
            }
            maxLength = Math.max(currentLength, maxLength);
        }
        return maxLength;
    }
}
